chore: export FPD without bbn chain

RafilxTenfen opened the pull request:
- Moved fpcli files to its own package to avoid multiple files as main package
- Add new cli command `fpcli export-finality-provider` that exports a finality provider from the database 
- Add new RPC on `fpd` client
```go
// SignMessageFromChainKey signs a message from the chain keyring.
SignMessageFromChainKey(ctx context.Context, in *SignMessageFromChainKeyRequest, opts ...grpc.CallOption) (*SignMessageFromChainKeyResponse, error)
```
> Used to sign exported finality provider

vitsalis commented on docs/phase-1/finality-provider.md:
```diff
@@ -0,0 +1,67 @@
+# Finality Provider
```
Let's wrap `.md` files to 80 char lines.

vitsalis replied:
Also, don't think this should be underneath a `phase-1` folder. The new functionality is about exporting finality provider information, doesn't have anything to do with the phase. Would also recommend renaming the file to `finality-provider-export.md`

RafilxTenfen replied:
done at https://github.com/babylonchain/finality-provider/pull/286/commits/7f9dadbaaf625ad9e3473a7195b2e1c96399444a

vitsalis commented on docs/phase-1/finality-provider.md:
```diff
@@ -0,0 +1,67 @@
+# Finality Provider
+
+Finality providers are responsible for voting at a finality round on top of CometBFT. Similar to any native PoS validator, a finality provider can receive voting power delegations from BTC stakers, and can earn commission from the staking rewards denominated in Babylon tokens.
```
```suggestion
Finality providers are responsible for voting at a finality round on top of CometBFT that is powered by BTC stake.
Similar to any native PoS validator, a finality provider can receive voting power delegations from BTC stakers, and through their voting power provide economic security to the underlying PoS protocol.

In this document,
we explore how someone can create a finality provider and export details about it in JSON format.
```

RafilxTenfen replied:
https://github.com/babylonchain/finality-provider/pull/286/commits/7f9dadbaaf625ad9e3473a7195b2e1c96399444a

vitsalis commented on docs/phase-1/finality-provider.md:
```diff
@@ -0,0 +1,67 @@
+
+## Creation & Export of Finality provider
+
+- Start by creating the finality provider config, this config is needed to define the `[dbconfig]` where it is set the path and database configurations
```
Do we need to use bullets here? Why not just a sequence of sentences?

vitsalis commented on docs/phase-1/finality-provider.md:
````diff
@@ -0,0 +1,67 @@
+- Start by creating the finality provider config, this config is needed to define the `[dbconfig]` where it is set the path and database configurations
+
+```shell
+$~ fpd init --home ./phase-1/fpd
````
```suggestion
$ fpd init --home ./phase-1/fpd
```

Same for others.

We should also show sample outputs of commands and explain what the CLI flags are.

vitsalis commented on docs/phase-1/finality-provider.md:
````diff
@@ -0,0 +1,67 @@
+$~ fpd init --home ./phase-1/fpd
+```
+
+- The finality provider need a private and public key on the consumer chain, in this case `babylon`, for creating a key run the following.:
````
```suggestion
The finality provider requires a key pair on the PoS chain it aims to secure. The chain is identified by its chain ID.
```

Also explanation on the other CLI flags.

vitsalis commented on docs/phase-1/finality-provider.md:
````diff
@@ -0,0 +1,67 @@
+$~ fpd keys add finality-provider --home ./phase-1/fpd --chain-id babylon-1 --key-name finality-provider
+```
+
+- Each finality provider needs to provide public randomness, for generating and managing such a thing the eots is used. Start by initiating the config file of eots with:
````
> for generating and managing such a thing the eots is used. 

Don't get what this means.

Below we explain how we set up the EOTS manager. We can write a sentence about what it is and why it is needed and explain that operations on it happen through the `eotsd` executable.

vitsalis commented on docs/phase-1/finality-provider.md:
````diff
@@ -0,0 +1,67 @@
+$~ eotsd init --home ./phase-1/eots
+```
+
+- In another terminal, start the eots manager daemon.
````
Why do we need the eots manager daemon to be operational? Ideally we should be able to do this in just CLI mode without the need to run daemons.

RafilxTenfen replied:
We could connect to the eots database directly from the fpcli, but that could be a bit messy and couple a lot the code packages

vitsalis commented on docs/phase-1/finality-provider.md:
````diff
@@ -0,0 +1,67 @@
+$~ eotsd start --home ./phase-1/eots
+```
+
+- On this phase no babylon or consumer chain is not running, so it is not possible to register the finality provider onchain, but for BTC delegations to choose your finality provider
````
I don't think we need to provide this information. This document is on creating the finality provider information and exporting it.

vitsalis commented on docs/phase-1/finality-provider.md:
````diff
@@ -0,0 +1,67 @@
+for staking their BTC, there is a need to get information about you, so the solution is to export the finality provider information with the following CLI:
+
+```shell
+$~ fpcli p1-export-finality-provider --home ./phase-1/fpd --key-name finality-provider --chain-id babylon-1 --commission 0.05 --moniker my-fp-nickname --identity anyIdentity --website www.my-public-available-website.com --security-contact email-for-questions@gmail.com --details 'other overall info'
````
This is a very long command, let's split it in multiple lines.

SebastianElvis commented on docs/finality-provider-export.md:
````diff
@@ -0,0 +1,142 @@
+  - `--details` any other optional detail information.
+
+```shell
+$ fpcli export-finality-provider --home ./export-fp/fpd --key-name finality-provider \
````
Tbh I feel the above sections are not relevant to this doc, whose focus is exporting a finality provider to JSON format.

So probably

- The above sections could be merged to `docs/finality-provider.md`
- The doc related to exporting command could be merged to `README.md`

Wdyt?

RafilxTenfen replied:
I think the idea for this `docs/finality-provider-export.md` file is to have instructions for finality providers on how they can create and export their data to register themselves without a bbn chain active

One thought is to receive this registration by pull requests in a github repository, but I believe it would be another repository

SebastianElvis commented on finality-provider/cmd/fpcli/daemon/export-fpd.go:
```diff
@@ -0,0 +1,170 @@
+			Usage: "The pass phrase used to encrypt the keys",
+			Value: defaultPassphrase,
+		},
+		cli.StringFlag{
```
I'm not sure if this field and the below ones are needed for exporting a fp. They can be fetced from DB rather than being provided by the user.

RafilxTenfen replied:
This CLI creates and export an fpd

Without the fpd daemon active, it creates a new FPD, stores it in the local db without sending any transactions to the chain, I will add a description to the CLI

RafilxTenfen replied:
done at https://github.com/babylonchain/finality-provider/pull/286/commits/e8ae68062ba0b4a76c8b8147cd4031ec4d1a310f

SebastianElvis commented on finality-provider/cmd/fpcli/daemon/export-fpd.go:
```diff
@@ -0,0 +1,170 @@
+		BabylonPk:            stored.ChainPk,
+		Description:          stored.Description,
+		Commission:           stored.Commission,
+		RegisteredEpoch:      0,
+		SlashedBabylonHeight: 0,
+		SlashedBtcHeight:     0,
```
These fields are stateful and out of control for the finality provider operator. So probably we are not using the right protobuf object. However, I'm not sure if we have one without these fields :sweat:

RafilxTenfen replied:
I agree, but the structure being exported here is the same one we can import on the genesis btc staking module

So, setting these stateful fields helps a little the import phase, where we don't need another protobuf finality provider structure to parse this

SebastianElvis reviewed (approved):
Lgtm! Would be great to get feedback from @gitferry as well :+1:

RafilxTenfen commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,142 @@
+
+## Requirements
+
+- [eotsd](https://github.com/babylonchain/finality-provider)
+- [fpd](https://github.com/babylonchain/finality-provider)
```
what do you think about linking it ? https://github.com/babylonchain/finality-provider/pull/286/commits/ef56bc4a456fa04a0b0252ebde483b2a8c10e5df

RafilxTenfen commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,142 @@
+# Finality Provider Export
+
+Finality providers are responsible for voting at a finality round on top of CometBFT that is powered by BTC stake.
+Similar to any native PoS validator, a finality provider can receive voting power delegations from BTC stakers, and
```
done at https://github.com/babylonchain/finality-provider/pull/286/commits/9917b4e1cb4b4e70a06d887f89be0ef1d3983852

RafilxTenfen commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,142 @@
+In this document,
+we explore how someone can create a finality provider and export details about it in JSON format.
+
+## Requirements
```
done at https://github.com/babylonchain/finality-provider/pull/286/commits/6e5ef32bc2451fa38ca548e6645a4d4323a52148

RafilxTenfen commented on finality-provider/cmd/fpcli/daemon/export-fpd.go:
agree, is already inside `finality-provider`

RafilxTenfen commented on finality-provider/cmd/fpcli/daemon/export-fpd.go:
```diff
@@ -0,0 +1,174 @@
+	Usage:     "Create, stores, and exports one finality provider.",
+	Description: `Connects with the EOTS manager defined in config, creates a new
+key pair formatted by BIP-340 (Schnorr Signatures), generates the master public
+randomness pair, stores the finality provider and export it by printing the json
```
done at https://github.com/babylonchain/finality-provider/pull/286/commits/9931d4cf2c5cbb82dfbc13c38aa4bc494a5b579f

RafilxTenfen commented on finality-provider/cmd/fpcli/daemon/export-fpd.go:
```diff
@@ -0,0 +1,174 @@
+var ExportFinalityProvider = cli.Command{
+	Name:      "export-finality-provider",
+	ShortName: "exfpd",
+	Usage:     "Create, stores, and exports one finality provider.",
```
done at https://github.com/babylonchain/finality-provider/pull/286/commits/9931d4cf2c5cbb82dfbc13c38aa4bc494a5b579f

RafilxTenfen commented on finality-provider/cmd/fpcli/daemon/export-fpd.go:
```diff
@@ -0,0 +1,174 @@
+
+var ExportFinalityProvider = cli.Command{
+	Name:      "export-finality-provider",
+	ShortName: "exfpd",
```
done at https://github.com/babylonchain/finality-provider/pull/286/commits/9931d4cf2c5cbb82dfbc13c38aa4bc494a5b579f

gitferry reviewed (commented):
After thinking more about this, I feel like the `export-finality-provider` cmd introduced here actually creates a finality provider and prints out the infomation. The creating part seems unnecessary. What if the user loses the info and wants to export it again? Maybe we can use the old `create-finality-provider` cmd to create the finality provider and export it by specifying the `btc-pk`? e.g., `fpcli export-finality-provider --btc-pk xxxxxx path/to/json`, which will not succeed if specified finality provider does not exist.

The current issue with `create-finality-provider` is that it requires `fpd` to run and `fpd start` will start all the existing finality providers, which are not needed if we only need to export a created one. Maybe we just need to introduce something like a `pure` mode to just start `fpd` without running existing finality providers. Nevertheless, I think relying the running of `fpd` is necessary because it is the `fpd` that controls the db and we should not let `fpcli` to connect to db directly. This way, validators can just follow the previous guidance to create finality providers and export it with an additional cmd. Wdyt? cc @vitsalis @SebastianElvis 

So, I propose the following changes based on this PR (or we can make it in a new PR if you feel cleaner):
1. remove the creation part of the current `export-finality-provider` and allow export by specifying `btc-pk` and sending a rpc request to `fpd` to get the data, similar to `fpcli finality-provider-info`.
2. introduce a `pure` mode to `fpd start` to allow the daemon to start without initiating any existing finality providers
3. the export part of the `docs/finality-provider-export.md` can be moved to `docs/finality-provider.md` with some changes about the introducing of `pure` mode to `fpd start` when creating finality providers

Left some comments on the doc which probably can be ignored with the proposed changes are agreed:

RafilxTenfen commented:
> The current issue with `create-finality-provider` is that it requires `fpd` to run 

and also connects to the bbn chain to register the fp by the transaction, this command `export-finality-provider` could even be removed after the phase-1 sunset, the idea was to make it simple for the finality provider to generate the needed information for registration. 

In a better scenario, we would not even need to run `eotsd`, this cli would connect directly to the eotsd db, but doing that would mess up a lot the code

RafilxTenfen commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,152 @@
+- `fpd`
+- `fpcli`
+
+### Setup EOTS
```
done at https://github.com/babylonchain/finality-provider/pull/286/commits/503d663df53b4c991a0670dc21a7d0f43704265e

RafilxTenfen commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,152 @@
+So, to export one finality provider it is needed to register a new key.
+
+To initialize the eots work directory, run `eotsd init`. It creates in the default
+home location, unless `--home` flag is specified.
```
done at https://github.com/babylonchain/finality-provider/pull/286/commits/503d663df53b4c991a0670dc21a7d0f43704265e

RafilxTenfen commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,152 @@
+> Starts the eots process that can be turned down after the finality provider is exported
+> (run all commands of this file).
+
+### Setup fpd
```
done at https://github.com/babylonchain/finality-provider/pull/286/commits/503d663df53b4c991a0670dc21a7d0f43704265e

RafilxTenfen commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,152 @@
+
+- `eotsd`
+- `fpd`
+- `fpcli`
```
done at https://github.com/babylonchain/finality-provider/pull/286/commits/503d663df53b4c991a0670dc21a7d0f43704265e

RafilxTenfen commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,152 @@
+export his information, there is no need to connect to babylon chain, just
+initialize the working directory for the finality provider config.
+To initialize the fpd work directory, run `fpd init`. It creates
+in the default home location, unless `--home` flag is specified.
```
done at https://github.com/babylonchain/finality-provider/pull/286/commits/503d663df53b4c991a0670dc21a7d0f43704265e

RafilxTenfen commented on docs/finality-provider-export.md:
````diff
@@ -0,0 +1,152 @@
+```
+
+> Creates one key pair identified by the key name `finality-provider`
+> Store safely the mnemonic generated
````
done at https://github.com/babylonchain/finality-provider/pull/286/commits/503d663df53b4c991a0670dc21a7d0f43704265e

RafilxTenfen commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,152 @@
+> Creates one key pair identified by the key name `finality-provider`
+> Store safely the mnemonic generated
+
+## Creation & Export of Finality provider
```
done at https://github.com/babylonchain/finality-provider/pull/286/commits/503d663df53b4c991a0670dc21a7d0f43704265e

gitferry commented:
> and also connects to the bbn chain to register the fp by the transaction, this command export-finality-provider could even be removed after the phase-1 sunset, the idea was to make it simple for the finality provider to generate the needed information for registration.

`create-finality-provider` does not do the registration. We have `register-finality-provider` to do that.

RafilxTenfen commented:
> `create-finality-provider` does not do the registration. We have `register-finality-provider` to do that.

true

So in this case, @gitferry  the idea is:
1. Run `eotsd init`
2. Run `eotsd start`
3. Run `fpd init`
4. Run `fpd start --no--bbn` (or something like that) to not connect to the client controller in the app
5. Run `fpcli create-finality-provider` generate the fp and output to the user the `btc_pk_hex` (key for getting the fp from the store)
6. Can turn down `eotsd`
7. Run `fpcli export-finality-provider --btc-pk-hex [btc_pk_hex]` to export the previously created finality provider by connecting to fpd process (not connecting directly to the kv database)
8. Can turn down `fpd`

I liked, this way we can maintain the export even after the phase-1

gitferry commented:
All steps are valid 👍. Just a note that we don't ask users to turn down daemons in the instructions but we can add a note in step 7 that this command does not rely on `eotsd` running

vitsalis commented:
We should also include instructions for safe-keeping and not losing the finality provider key.

gitferry commented on finality-provider/cmd/fpcli/daemon/export.go:
```diff
@@ -0,0 +1,163 @@
+var ExportFinalityProvider = cli.Command{
+	Name:      "export-finality-provider",
+	ShortName: "exfp",
+	Usage:     "Creates, stores, and exports one finality provider.",
```
```suggestion
	Usage:     "It exports the finality provider by the given public key.",
```

RafilxTenfen replied:
good catch, my bad did not revised the descriptions :/

RafilxTenfen replied:
done at https://github.com/babylonchain/finality-provider/pull/286/commits/2adcd440abeaea18488d2edb6b93138dddbe776c

gitferry commented on finality-provider/cmd/fpcli/daemon/export.go:
```diff
@@ -0,0 +1,163 @@
+	Name:      "export-finality-provider",
+	ShortName: "exfp",
+	Usage:     "Creates, stores, and exports one finality provider.",
+	Description: `Connects with the EOTS manager defined in config, creates a new
+key pair formatted by BIP-340 (Schnorr Signatures), generates the master public
+randomness pair, stores the finality provider and exports it by printing the json
+structure on the stdout`,
```
```suggestion
	Description: `Fetches the finality provider from the database and exports it by printing the json
structure on the stdout`,
```

RafilxTenfen replied:
https://github.com/babylonchain/finality-provider/pull/286/commits/2adcd440abeaea18488d2edb6b93138dddbe776c

gitferry commented on finality-provider/cmd/fpcli/daemon/export.go:
```diff
@@ -0,0 +1,163 @@
+		cli.BoolFlag{
+			Name: signedFlag,
+			Usage: `Defines if needs to sign the exported finality provider,
+			if true, it is necessary to define keyring flags`,
```
```suggestion
			if true, it is necessary to define the key name flag`,
```

gitferry commented on finality-provider/cmd/fpcli/daemon/export.go:
```diff
@@ -0,0 +1,163 @@
+		printRespJSON(fp)
+		return nil
+	}
+
```
Do we need to check whether the key name is specified if `signed` is true?

RafilxTenfen replied:
by default if `--key-name` flag is not set, loads from config
If there is nothing on config, it errors out

RafilxTenfen replied:
also, when submitting this information for us, we have a way for the CI to verify if the signature matches
https://github.com/babylonchain/finality-provider/pull/290

Let me know what you think about this

gitferry commented on finality-provider/cmd/fpcli/daemon/export.go:
```diff
@@ -0,0 +1,163 @@
+	btcstktypes.FinalityProvider
+	// FpSigHex is the finality provider cosmos sdk chain key
+	// can be verified with the pub key in btcstktypes.FinalityProvider.BabylonPk
+	FpSigHex string `json:"fp_sig_hex"`
```
what do we need this sig for?

RafilxTenfen replied:
The idea is to make explicit that nobody modified the finality provider information once generated.
It is proof that the finality provider generated that information with those descriptions and more importantly, the pub keys and commission

RafilxTenfen replied:
When submitting this finality provider information for us, we have a way for the CI to verify if the signature matches
https://github.com/babylonchain/finality-provider/pull/290

Let me know what you think about this

gitferry commented on finality-provider/cmd/fpd/daemon/start.go:
```diff
@@ -1,21 +1,27 @@
 
-var startCommand = cli.Command{
+const (
+	noBabylonChainFlag = "no-bbn"
```
```suggestion
	noChainBackendFlag = "no-chain-backend"
```

RafilxTenfen replied:
done at https://github.com/babylonchain/finality-provider/pull/286/commits/0b19b133d19ccdd03743ed6ef827065d138981dd

gitferry commented on finality-provider/cmd/fpd/daemon/start.go:
```diff
@@ -38,6 +44,10 @@ var startCommand = cli.Command{
 		},
+		cli.BoolFlag{
+			Name:  noBabylonChainFlag,
+			Usage: "If specified, does not connect to client controller",
```
```suggestion
			Usage: "If specified, does not connect to the chain backend",
```

RafilxTenfen replied:
https://github.com/babylonchain/finality-provider/pull/286/commits/0b19b133d19ccdd03743ed6ef827065d138981dd

gitferry commented on finality-provider/cmd/fpd/daemon/start.go:
```diff
@@ -76,44 +83,81 @@ func start(ctx *cli.Context) error {
+	dbBackend walletdb.DB,
+) (*service.FinalityProviderApp, error) {
+	if ctx.Bool(noBabylonChainFlag) {
+		fpApp, err := service.NewFinalityProviderAppWithEotsFromConfig(cfg, dbBackend, logger)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create finality-provider with not bbn chain app: %v", err)
+		}
+
+		return fpApp, nil
+	}
+
 	fpApp, err := service.NewFinalityProviderAppFromConfig(cfg, dbBackend, logger)
```
why does it relate to Eots? Maybe we can just insert the chain controller as an additional input to the function

RafilxTenfen replied:
https://github.com/babylonchain/finality-provider/pull/286/commits/73ff4dd7ab0566921078b2742d4fb4ead3c3d58f

RafilxTenfen commented:
@gitferry I had not merged the `docs/finality-provider-export.md` into `docs/finality-provider.md` file, because it explains different perspectives, the export one setups eotsd and fpd just for generating information that finality providers will use to submit their information 
In a latter phase, we can have another set of instructions just for they to turn fpd and eotsd up when the chain starts

gitferry reviewed (commented):
Thanks for the change. Great work! Some small comments:

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,211 @@
+
+> Creates the config file and two directories for logs and data.
+
+After the proper configuration, start the etos daemon with the command `eotsd start`.
```
```suggestion
After the proper configuration, start the eots daemon with the command `eotsd start`.
```

RafilxTenfen replied:
https://github.com/babylonchain/finality-provider/pull/286/commits/c24afebdef4c435f69b345989989efda404ff596

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,211 @@
+### Setup Finality Provider Daemon
+
+The finality provider daemon (fpd) is responsible for polling consumer chain blocks
+and providing finality signatures if it has voting power. To initialize the `fpd`
```
```suggestion
and providing finality signatures if it is in the active finality provider set.
To initialize the `fpd`
```

RafilxTenfen replied:
https://github.com/babylonchain/finality-provider/pull/286/commits/5d51cd265c9fd1ba3096f6365f223e61af64881b

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,211 @@
+, one of `[os, file, pass, kwallet]`.
+
+After the setup of the configuration and key to be used, start the `fpd` daemon with the command `fpd start` with no babylon chain `--no-bbn`.
+The home golder should point to the work directory previously created with `fpd init`
```
The above need to be wrapped to 80 char lines

RafilxTenfen replied:
https://github.com/babylonchain/finality-provider/pull/286/commits/deb3c487d0427149bfe20f99f0f04fc7dff1f1b8
https://github.com/babylonchain/finality-provider/pull/286/commits/5d51cd265c9fd1ba3096f6365f223e61af64881b

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,211 @@
+> For production enviroments, make sure to select a proper [backend keyring](https://docs.cosmos.network/v0.45/run-node/keyring.html#available-backends-for-the-keyring)
+, one of `[os, file, pass, kwallet]`.
+
+After the setup of the configuration and key to be used, start the `fpd` daemon with the command `fpd start` with no babylon chain `--no-bbn`.
```
Should we say why no Babylon chain is needed?

RafilxTenfen replied:
https://github.com/babylonchain/finality-provider/pull/286/commits/1892c9ae733df3e0c17cbc8be3c84b8cdc5f0a67

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,211 @@
+## Create a Finality Provider
+
+After the setup and start of `eots` and `fdp`, to interact with the `fpd`
+daemon, it is used the `fpcli`. To create the finality provider, run
```
```suggestion
daemon, you can use the `fpcli` utility. To create the finality provider, run
```

RafilxTenfen replied:
https://github.com/babylonchain/finality-provider/pull/286/commits/0a0e83ee02d41ccf9607118a7732af5274e6b257

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,211 @@
+daemon, it is used the `fpcli`. To create the finality provider, run
+`fpcli create-finality-provider`.
+
+> Obs.: This command does not send a transaction to babylon chain `babylond tx btcstaking create-finality-provider`.
```
I don't get this line, why do we specify the babylon cli here?

vitsalis commented on docs/finality-provider-export.md:
````diff
@@ -0,0 +1,211 @@
+--signed
+```
+
+The expected result is a finality with 5% comission exported:
````
Why do we specify the commission here? Someone might think that we always output 5% commission. I would remove it.

RafilxTenfen replied:
removed at https://github.com/babylonchain/finality-provider/pull/286/commits/0c72c3bd18cec397af881f69eb228fcb5d5a1831

RafilxTenfen commented on finality-provider/cmd/fpcli/daemon/export.go:
```diff
@@ -0,0 +1,163 @@
+)
+
+// FinalityProviderSigned wrap the finality provider by adding the
+// signed finality probider by the bbn pub key as hex
```
done at https://github.com/babylonchain/finality-provider/pull/286/commits/b100888615e074809063e7789252032444bd7876

RafilxTenfen commented:
![image](https://github.com/babylonchain/finality-provider/assets/17556614/5a510281-7fa2-4d9c-a3e6-13e8263b4a71)

RafilxTenfen commented on docs/finality-provider-export.md:
````diff
@@ -0,0 +1,249 @@
+> Keep the database folders of `eots` and `fpd` securely stored as it contains
+information that will be used latter for running the finality provider.
+
+## Security considerations
+
+All the generated files and keyring must be stored in a secure place for latter usage.
+If any keys are lost, BTC or backend chain the finality provider will not be able
+to start the chain with BTC delegations prior genesis block and all the BTC locked
+in the finality provider that lost his keys will not be providing security
+and by consequent, not earn any rewards.
+
+The recommendation is to safely store the entire `./export-fp` directory
+that compress occupies roughly 5kb. To compress, run the following command:
+
+```shell
+tar czf backup-exported-fpd.tar.gz ./export-fp
+```
+
+⚠ Keep the compressed directory stored securely.
````
New section that explains what happens if the keys are lost, should I mention that no BTC would be lost even if the finality provider lost the keys? 
@vitsalis

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,249 @@
+
+We need the EOTS manager daemon (`eotsd`) to create and manage EOTS keys for each
+finality provider. To initialize the `eotsd` work directory, run `eotsd init`.
+It creates the default home location, unless `--home` flag is specified.
```
What's the default home location?

RafilxTenfen replied:
$HOME/.eotsd

vitsalis commented on docs/finality-provider-export.md:
````diff
@@ -0,0 +1,249 @@
+2024-04-17T16:52:10.554061Z     info    EOTS Manager Daemon is fully active!
+```
+
+> Starts the eots process that can be turned down after the finality provider
+> is exported (run all commands of this file).
````
```suggestion
> The eots process can be turned down after this guide is completed
> if the goal is to just export the finality provider and not connect it to a PoS chain.
```

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,249 @@
+
+### Setup Finality Provider Daemon
+
+The finality provider daemon (fpd) is responsible for polling consumer chain blocks
```
```suggestion
The finality provider daemon (`fpd`) is responsible for polling consumer chain blocks
```

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,249 @@
+> The added key will be used to create the proof-of-possession (pop)
+> of the finality provider.
+> For production enviroments, make sure to select a proper
+[backend keyring](https://docs.cosmos.network/v0.45/run-node/keyring.html#available-backends-for-the-keyring)
```
```suggestion
[keyring backend](https://docs.cosmos.network/v0.45/run-node/keyring.html#available-backends-for-the-keyring)
```

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,249 @@
+> of the finality provider.
+> For production enviroments, make sure to select a proper
+[backend keyring](https://docs.cosmos.network/v0.45/run-node/keyring.html#available-backends-for-the-keyring)
+, one of `[os, file, pass, kwallet]`.
```
```suggestion
, i.e. one of `[os, file, pass, kwallet]`.
```

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,249 @@
+, one of `[os, file, pass, kwallet]`.
+
+After the setup of the configuration and key to be used, start the `fpd` daemon
+with the command `fpd start` with no chain backend `--no-chain-backend`.
+No chain backend is needed to create and export the finality provider information.
```
```suggestion
with the command `fpd start`. If you want to just export the finality provider info without
connecting to a PoS chain, then you can use the flag `--no-chain-backend`.
```

Don't think `--no-chain-backend` is a pre-requisite for exporting, right?

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,249 @@
+
+## Create a Finality Provider
+
+After the setup and start of `eots` and `fdp`, to interact with the `fpd`
```
```suggestion
After the setup and start of `eots` and `fpd`, to interact with the `fpd`
```

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,249 @@
+daemon, you can use the `fpcli` utility. To create the finality provider,
+run `fpcli create-finality-provider`.
+
+> Obs.: This command does not send a transaction to babylon chain.
```
```suggestion
> Obs.: This command only registers the finality provider in the local `fpd` database and not to the Babylon chain, meaning that it can be ran without a connection to it.
```

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,249 @@
+
+## Export a Finality Provider
+
+Finally, after the creation of the finality provider, to export the finality
```
```suggestion
Finally, after the creation of the finality provider in the local db, to export the finality
```

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,249 @@
+
+Finally, after the creation of the finality provider, to export the finality
+provider, run `fpcli export-finality-provider`. This command connects with the
+`fpd` daemon to load the finality provider previously created using the flag `--btc-pk`.
```
```suggestion
`fpd` daemon to load the finality provider previously created using the flag `--btc-pk`
to specify the BTC public key that acts as the unique identifier for it.
```

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,249 @@
+
+- `--btc-pk` the hex string of the BTC public key.
+- `--signed` signs the finality provider as a proof of a untempered exported data.
+- `--key-name` mandatory as it identifies the name for the key to use on fpd creation.
```
Why is `key-name` and not `btc-pk` the mandatory flag? I would expect that we can find the finality provider just with the `--btc-pk`.

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,249 @@
+provider, run `fpcli export-finality-provider`. This command connects with the
+`fpd` daemon to load the finality provider previously created using the flag `--btc-pk`.
+
+This command also has several flag options.:
```
Shouldn't we specify `--keyring-backend` as well here?

vitsalis commented on docs/finality-provider-export.md:
```diff
@@ -0,0 +1,249 @@
+
+> The eots manager (`eotsd start`) can be turned down.
+> The fpd daemon (`fpd start`) can be turned down.
+> Keep the database folders of `eots` and `fpd` securely stored as it contains
+information that will be used latter for running the finality provider.
```
```suggestion
> In the case that there is no Babylon chain running, but you intend to connect to it
once it is up, you should keep the database folders of `eotsd` and `fpd` securely stored
as they are necessary for retrieving your finality provider and registering it with the chain.
```

RafilxTenfen replied:
file removed

RafilxTenfen commented:
As discussed, removed flag for start `fpd` without bbn chain 
Removed specific .md file for export CLI and add an export example to `finality-provider.md`

vitsalis commented on finality-provider/cmd/fpcli/daemon/export.go:
```diff
@@ -0,0 +1,161 @@
+		},
+		cli.BoolFlag{
+			Name: signedFlag,
+			Usage: `Defines if needs to sign the exported finality provider,
```
```suggestion
			Usage: `Specify if the exported finality provider information should be signed,
```

vitsalis commented on finality-provider/cmd/fpcli/daemon/export.go:
```diff
@@ -0,0 +1,161 @@
+		return fmt.Errorf("invalid fp btc pk hex %s: %w", fpBtcPkHex, err)
+	}
+
+	fpInfoResp, err := client.QueryFinalityProviderInfo(context.Background(), fpPk)
```
```suggestion
	fpInfoResp, err := client.QueryFinalityProviderInfo(ctx, fpPk)
```

RafilxTenfen replied:
the ctx in this case is `ctx *cli.Context` not able to parse it to `ctx context.Context`

vitsalis commented on finality-provider/cmd/fpcli/daemon/export.go:
```diff
@@ -0,0 +1,161 @@
+	}
+
+	resp, err := client.SignMessageFromChainKey(
+		context.Background(),
```
```suggestion
		ctx,
```

RafilxTenfen replied:
the ctx in this case is ctx *cli.Context not able to parse it to ctx context.Context

vitsalis commented on finality-provider/service/app.go:
```diff
@@ -345,67 +345,106 @@ func (app *FinalityProviderApp) CreateFinalityProvider(
+
+// loadChainKeyring checks the keyring by loading or creating a chain key.
+func (app *FinalityProviderApp) loadChainKeyring(
+	keyName, passPhrase, hdPath string,
```
Is a passphrase and hdpath required? What happens if they are not set?

RafilxTenfen replied:
it accepts it as empty strings, so they are not required

vitsalis reviewed (approved):
Nice work! Minor comments

RafilxTenfen commented on docs/finality-provider.md:
```diff
@@ -215,3 +215,50 @@ fpcli list-finality-providers
+- `--daemon-address` the RPC server address of `fpd` daemon.
+- `--signed` signs the finality provider with chain key as a proof of a
+untempered exported data.
+- `--key-name` mandatory as it identifies the name for the key to use on fpd creation.
```
my bad, key-name flag is only necessary if the export should be signed and there is no key-name set on config
I will updated the description

RafilxTenfen commented on docs/finality-provider.md:
```diff
@@ -215,3 +215,50 @@ fpcli list-finality-providers
+- `--key-name` mandatory as it identifies the name for the key to use on fpd creation.
+- `--home` flag needs to be the same as used in the `fpd init` command,
+to load the configuration.
+- `--passphrase` the password used to encrypt the key.
```
Since the `fpcli` does not load the keys, the App on `fpd` uses the keyring backend from config
